### DISCUSSION TITLE
Implementing `rules()` on a top level object will disable underlying validation

### DISCUSSION THREAD

**Discussion by arondeparon** [spatie/laravel-data#132] [Q&A]
I am not sure if this intended, a bug or just my ignorance, but I have noticed the following:

When I add a `rules()` method to a top level object that contains a `DataCollection` property which also implements a `rules()` method, the underlying rules method will be called.

Example:

```php
class Listings extends Data
{
    public function __construct(
        /** @var \App\Dto\Listing[] */
        public DataCollection $listings,
    )
    {
    }

    // removing this method will allow the underlying object to be validated.
    public static function rules(): array
    {
        return [
            'listings' => ['required', 'array'],
        ];
    }
}
```

```php
class Listing extends Data
{
    public static function rules(): array
    {
        // will not be called when injecting `Listings` in a controller. 
    }
}
```

Looking at the code, this seems to be caused by [this line](https://github.com/spatie/laravel-data/blob/main/src/Resolvers/DataValidationRulesResolver.php#L29):

```php
            ->reject(fn (DataProperty $property) => array_key_exists($property->name(), $overWrittenRules) || ! $property->shouldValidateProperty())
```

It seems like mentioning an underlying property will basically "overwrite" the validation and ignore anything that happens after that.

**So, the question: how would I implement this pattern while making sure that every validator is called?**

**Comment by rubenvanassche** [comment]
Hi @ArondeParon,

Could you provide me a test which shows where this is going wrong? We have a test for nested data collections with overwritten rules : https://github.com/spatie/laravel-data/blob/main/tests/Resolvers/DataValidationRulesResolverTest.php#L74-L90 and it seems to be working fine to me.

**Reply by benperiton** [reply]
I've just come across this as well, an Order DTO with a collection of OrderLine, if the Order has rules to make sure that it is required and an array, then the rules for OrderLine don't get run.

```php
class OrderData extends Data
{
    public function __construct(
        public readonly string $name,
        #[DataCollectionOf(OrderLineData::class)]
        public readonly DataCollection $lines,
    ) {
    }
    
    public static function rules(): array
    {
        return [
            'name' => ['required', 'max:20'],
            'lines' => ['required', 'array'],
        ];
    }
}

class OrderLineData extends Data
{
    public function __construct(
        public readonly int $qty,
    ) {}

    public static function rules(): array
    {
        return [
            'qty' => ['required', 'int'],
        ];
    }
}
```

If I create the OrderData object with some line items missing the 'qty' field, then the rules are not run and instead a TypeError is thrown
````
OrderLineData::__construct(): Argument #1 ($qty) must be of type int, null given
````

If I remove the `'lines' => ['required', 'array'],` line, then it runs the validation rules in OrderLineData

To me it would make sense that the OrderData can validate that there are actually lines passed in, and then the OrderLineData can validate that line data is correct, at the moment it seems only one or the other can work.

**Reply by rubenvanassche** [reply]
Thanks for the example, I've added a fix in the latest release!

**Reply by benperiton** [reply]
So that totally works, except for messages on the collection - but that seems that its because it is never called.

Perhaps need a new DataClassValidationMessagesResolver ?

**Reply by rubenvanassche** [reply]
You're always welcome to PR such a thing!
